### PR TITLE
Priority order broken when combined with dependencies that have failures

### LINKED ISSUE #1334: parallel="classes" does not work if class extends abstract class ,but works fine parallel="tests"
### Expected behavior
should work

### Actual behavior
does not work. stopped at abstract class method.

### Is the issue reproductible on runner?

- [ ] Shell
- [X] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [ ] IntelliJ
- [X] NetBeans

====testng.xml====

```xml
<?xml version='1.0' encoding='UTF-8' ?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="com.test_e2e_jar_0.0.1-SNAPSHOT" **parallel="classes"** thread-count="3" configfailurepolicy="continue">

    <!--
    see examples at http://testng.org/doc/documentation-main.html#testng-xml

    <suite-files>
        <suite-file path="./junit-suite.xml" />
    </suite-files>

    <test name="TimeOut">
        <classes>
            <class name="test.timeout.TimeOutTest" />
            <class name="test.timeout.TimeOutFromXmlTest"/>
            <class name="test.timeout.TimeOutThreadLocalSampleTest"/>
        </classes>
    </test>
    -->
    
    <test name="AutoNumberCategoryTest">
        <classes>
            <class name="com.test.e2e.testcase.common.AutoNumberCategoryTest" />
            <class name="com.test.e2e.testcase.common.AutoNumberingTest" />
            <class name="com.test.e2e.testcase.common.CodeTypeTest" />
            <class name="com.test.e2e.testcase.common.ContactTypeTest" />
            <class name="com.test.e2e.testcase.common.PropertyConditionTest" />
            <class name="com.test.e2e.testcase.common.PropertyTypeTest" />
            <class name="com.test.e2e.testcase.common.UOMTest" />
            <class name="com.test.e2e.testcase.common.UOMConversionTest" />
        </classes>
    </test>
</suite>
```
parallel mode works fine if specify parallel="tests" in testng.xml:

```xml
<?xml version='1.0' encoding='UTF-8' ?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
<suite name="com_test_e2e_jar_0.0.1-SNAPSHOT" **## parallel="tests"** thread-count="3" configfailurepolicy="continue">

    <!--
    see examples at http://testng.org/doc/documentation-main.html#testng-xml

    <suite-files>
        <suite-file path="./junit-suite.xml" />
    </suite-files>

    <test name="TimeOut">
        <classes>
            <class name="test.timeout.TimeOutTest" />
            <class name="test.timeout.TimeOutFromXmlTest"/>
            <class name="test.timeout.TimeOutThreadLocalSampleTest"/>
        </classes>
    </test>
    -->
    
    <test name="AutoNumberCategoryTest">
        <classes>
            <class name="com.test.e2e.testcase.common.AutoNumberCategoryTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>      
        </classes>
      
    </test>
    <test name="AutoNumberingTest">
        <classes>
            <class name="com.test.e2e.testcase.common.AutoNumberingTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>
        </classes>

    </test>
    
    <!-- <test name="CodeReferencesTest">
        <classes>
            <class name="com.test.e2e.testcase.common.CodeReferencesTest" >
               <methods>
                    <include name="test0060_deleteAction" />
                </methods>
            </class>      
        </classes>
    </test>-->
    
    <test name="CodeTypeTest">
        <classes>
            <class name="com.test.e2e.testcase.common.CodeTypeTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>      
        </classes>
    </test>

    
     <test name="ContactTypeTest">
        <classes>
            <class name="com.test.e2e.testcase.common.ContactTypeTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>      
        </classes>
    </test>

    
<!--    <test name="LicenseTypeTest">
        <classes>
            <class name="com.test.e2e.testcase.common.LicenseTypeTest" >
            </class>      
        </classes>
    </test>-->

    
    <test name="PropertyConditionTest">
        <classes>
            <class name="com.test.e2e.testcase.common.PropertyConditionTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>      
        </classes>
    </test>

    
    <test name="PropertyTypeTest">
        <classes>
            <class name="com.test.e2e.testcase.common.PropertyTypeTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>      
        </classes>
    </test>

    
    <test name="UOMTest">
        <classes>
            <class name="com.test.e2e.testcase.common.UOMTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>      
        </classes>
    </test>

    
    <test name="UOMConversionTest">
        <classes>
            <class name="com.test.e2e.testcase.common.UOMConversionTest" >
<!--                <methods>
                    <include name="test0060_deleteAction" />
                </methods>-->
            </class>      
        </classes>
    </test>
</suite>
```

### PR DESCRIPTION
**Pull request by krmahadevan**
Fixes #1334 

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

### CODE REVIEW

**Comment by krmahadevan** [comment]
ping @juherr

File: src/main/java/org/testng/internal/TestMethodWorker.java [hunk 1]
```diff
@@ -53,13 +55,22 @@ public TestMethodWorker(IInvoker invoker,
...
   {
     m_invoker = invoker;
-    m_methodInstances = testMethods;
     m_suite = suite;
     m_parameters = parameters;
     m_groupMethods = groupMethods;
     m_classMethodMap = classMethodMap;
     m_testContext = testContext;
     m_listeners = listeners;
+    List<IMethodInstance> localMethodInstances = Lists.newArrayList(testMethods);
```
Comment (juherr) [inline src/main/java/org/testng/internal/TestMethodWorker.java]: The order is supposed to be managed by the graph when we add edges.

At the opposite, I'm trying to remove all other locations where an ordering is done: https://github.com/cbeust/testng/pull/1352

File: src/test/java/test/priority/PriorityTest.java [hunk 2]
```diff
@@ -46,4 +49,15 @@ public void priorityInSequential2() {
...
   }
+
+  @Test(description = "GITHUB #1334: Order by priority gets messed up when there are failures and dependsOnMethods")
+  public void priorityWithDependencyAndFailures() {
+    TestNG tng = create(SampleTest01.class, SampleTest02.class);
+    tng.setParallel(XmlSuite.ParallelMode.CLASSES);
+    tng.run();
+    List<String> sampleTest01Methods = Arrays.asList("test0010_createAction", "test0030_advancedSearch",
+        "test0060_deleteAction");
+    Assert.assertEquals(TestListener.methods.get(SampleTest01.class), sampleTest01Methods);
```
Comment (juherr) [inline src/test/java/test/priority/PriorityTest.java]: Pls, use `InvokedMethodNameListener`

File: CHANGES.txt [hunk 3]
```diff
@@ -1,4 +1,5 @@
 ﻿Current
+Fixed: GITHUB-1334: Priority order broken when combined with dependencies that have failures (Krishnan Mahadevan)
```
Comment (juherr) [inline CHANGES.txt]: Don't forget to add "Close #xxx" in the commit message. It will close the issue automatically.

**Review by juherr — CHANGES_REQUESTED** [review]
I'm not feeling good the fix. I think it is not fixing the root issue.

File: src/main/java/org/testng/internal/TestMethodWorker.java (hunk 1 above)
Comment (krmahadevan) [inline src/main/java/org/testng/internal/TestMethodWorker.java]: @juherr - I already tried applying changes part of #1352 before submitting this PR. It does not solve the problem.

File: src/test/java/test/priority/PriorityTest.java (hunk 2 above)
Comment (krmahadevan) [inline src/test/java/test/priority/PriorityTest.java]: Fixed.

File: CHANGES.txt (hunk 3 above)
Comment (krmahadevan) [inline CHANGES.txt]: Added that in the additional commit.

File: src/main/java/org/testng/internal/TestMethodWorker.java (hunk 1 above)
Comment (juherr) [inline src/main/java/org/testng/internal/TestMethodWorker.java]: In fact, what I mean is: don't try to sort method outside the graph.
BTW, I propose we add a comment which explains the fix is a workaround for the graph issue and we create a new issue for the graph one with a link to the current pull request.
WDYT?

Comment (krmahadevan) [inline src/main/java/org/testng/internal/TestMethodWorker.java]: @juherr - I think I may have fixed the ordering properly as part of [this](https://github.com/cbeust/testng/pull/1345/commits/f67cd34170d0daacce7df03a3a7c10a727563656) commit in PR #1345 

Can we first get #1345 merged, and after that I will test out this without sorting and see if that solves the problem ?

Comment (cbeust) [inline src/main/java/org/testng/internal/TestMethodWorker.java]: @juherr I fully support the idea of linking to Github issues in the source to explain why this code is here.

Comment (krmahadevan) [inline src/main/java/org/testng/internal/TestMethodWorker.java]: @cbeust - Sure I will do that. But I would like to see if #1345 fixes this problem without needing this PR. So I am waiting for #1345 to be merged before moving forward with this PR.

**Comment by krmahadevan** [comment]
@juherr - This PR merely adds tests for #1334 (since my PR #1361 fixed this problem). Please help take a look

**Comment by krmahadevan** [comment]
@cbeust - Can you please help get this merged. This PR contains only tests ? (Trying to wrap up all my existing PRs before starting work on new stuff)

**Comment by cbeust** [comment]
@juherr had a few doubts about this PR. Julien, you ok with merging this?

**Comment by krmahadevan** [comment]
@cbeust - Yes that's correct. But after #1361 got merged, this PR didn't require any code change. This PR just adds tests to re-iterate that this issue is really fixed. If you look at the diff, you will see that all the changes are under `src/test/java` only.

**Pull request merged**